src/cmd/shared.go

```diff
@@ -35,16 +47,39 @@ func validateBooleanArgument(arg string) error {
 	return nil
 }
 
+func validateBooleanArgumentFunc(arg string) func() error {
+	return func() error {
+		return validateBooleanArgument(arg)
+	}
+}
+
 func validateMinArgs(args []string, min int) error {
 	if len(args) < min {
 		return errors.New("Too few arguments")
 	}
 	return nil
 }
 
+func validateMinArgsFunc(args []string, min int) func() error {
+	return func() error {
+		return validateMinArgs(args, min)
+	}
+}
+
 func validateMaxArgs(args []string, max int) error {
 	if len(args) > max {
 		return errors.New("Too many arguments")
 	}
 	return nil
 }
+
+func validateMaxArgsFunc(args []string, max int) func() error {
+	return func() error {
+		return validateMaxArgs(args, max)
+	}
+}
+
+func validateIsConfigured() error {
+	prompt.EnsureIsConfigured()
+	return nil
+}
```

Comment:
@kevgo Just something I was thinking a bit about: if this type of function becomes a common pattern, I think we could create a nice helper:
```go
func noError(fn func()) {
  fn()
  return nil
}
```
And then we would use it as
```go
return util.FirstError(
  validateMaxArgsFunc(args, 0),
  git.ValidateIsRepository,
  util.noError(prompt.EnsureIsConfigured),
  git.ValidateIsOnline,
  drivers.ValidateHasDriver,
)
```

If you like that I could add that to this PR now.

Reply:
That would be the ultimate expression of leaning on the type system here. With this implemented, we now say "every function must return an error here". If a function doesn't return an error, the developer must "flag" this method as "its okay in this particular case". At this point, the developer takes responsibility that this hack around the type restriction is fine.

I still think a more flexible, dynamically typed API as I did initially would work here as well, and lead to even simpler and easier to understand code. But I'm okay with going the type system route here for the sake of experimentation. 

Btw, I think your `noError` func should return an `error`.  :)

Reply:
Sorry, it should probably be called `noErrorFunc`. The part I don't like about the more flexible API is the validation has to happen at run time instead of at compile time (which I think is the main benefit of the type system). 

And this method to me isn't a "hack" but is the way a developer inserts a function into the chain that won't error without having to make a function that could never return an error have that signature.

Reply:
I like that `util.noError` is short and semantically describes what we do here (along the lines of "btw this func doesn't return an error and that's okay here"). 

I meant "hack" as bypassing the documented requirement that all functions return errors here. By pretending to the type system that an error will be returned here while fully knowing this will never actually happen. "Hack" is too strong of a word here, but I don't have a better one. The point is it "tricks" the type system into accepting a function that it shouldn't be accepting, because that function doesn't fit into the documented type restriction here but we want to use it anyways.